### ISSUE TITLE
MsgBox mock/wrapper

### ISSUE THREAD

**Issue by retailcoder** [rubberduck-vba/Rubberduck#1228]
`MsgBox` is a common dependency that can crawl just about everywhere in a code base, and thwart unit testing (because it requires clicking, which defeats automated tests).

Let's pack a configurable message box wrapper with Rubberduck, that unit tests can use to predetermine the output (as one would with a mock), and that production code can use to actually display a `MsgBox`.

Something like this perhaps, in an `IMsgBox` class module:

```
VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = "IMsgBox"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = True
Attribute VB_Exposed = False
Option Explicit

Private Type TMsgBox
    IsMock As Boolean
    Result As VbMsgBoxResult
End Type

Private this As TMsgBox

Public Sub Setup(ByVal value As VbMsgBoxResult)
Attribute Setup.VB_Description = "Configures mock message box result, for testing purposes."
    this.IsMock = True
    this.Result = value
End Sub

Public Sub Release()
Attribute Release.VB_Description = "Releases an instance, which then no longer acts as a test mock."
    this.IsMock = False
End Sub

Public Property Get Result() As VbMsgBoxResult
    Result = this.Result
End Property

Public Function Show(ByVal prompt As String, Optional ByVal style As VbMsgBoxStyle, Optional ByVal title As String, Optional ByVal helpFile As Variant, Optional ByVal context As Variant) As VbMsgBoxResult
Attribute Show.VB_Description = "Displays the message box and returns the result. If Setup was called, only returns the setup result."
Attribute Show.VB_UserMemId = 0

    If Not this.IsMock Then
        If title = vbNullString Then title = Application.Name
        this.Result = MsgBox(prompt, style, title, helpFile, context)
    End If

    Show = this.Result

End Function
```

Thoughts?

Because of the `PredeclaredId` and `UserMemId = 0` attributes, in-code usage could look like this:

```
IMsgBox.Show "Hi!"
```

Or:

```
If IMsgBox("Confirm?", vbYesNo) = vbYes Then
```

This would be illegal though (compile error / invalid use of property - go figure):

```
IMsgBox "Hi!"
```

---

The benefit is that it makes a fairly easy to use wrapper/substitute for the commonly-used `MsgBox` built-in function, and code that uses the _default instance_ would still be configurable globally, so unit tests could do this:

```
IMsgBox.Setup vbYes
```

And then every single call to `IMsgBox.Show` (implicit or explicit) would return `vbYes`.

Ideally, user code wouldn't be so tightly coupled with the default instance though - but providing it does feel right. User code would ideally work with a given instance of `IMsgBox`, which test methods could "inject" with a given configuration.

**Comment by ThunderFrame** [comment]
Nice idea. I've shadowed the MsgBox and InputBox functions in the past, overriding the dialogs with UserForms, so I get more control over the behavior and look of the message.

There is a special case of MsgBox in Access, where you can achieve certain amounts of bold formatting in a message, by calling MsgBox inside an Eval statement, and including tactically placed `@` symbols in the message. A throwback to Access Basic.

**Comment by ThunderFrame** [comment]
But why the need for a static class? Couldn't we just shadow the VBA.MsgBox with our own? I guess we'd have to check if MsgBox was already shadowed, and maybe use a #If #Else block to shadow the shadow?

**Comment by retailcoder** [comment]
Because KISS

**Comment by retailcoder** [comment]
The idea is to remove explicit and direct `MsgBox` calls from usercode. And yeah, a similar wrapper would be needed for `InputBox`, with a `Setup` method that presets a determined value.

> "But what if my method is calling `InputBox` five times to retrieve five different values?"

Then you have five different concerns that each belong in their own method, or more accurately, you probably need to wrap some values into a type, and refactor things a bit.

Writing truly *testable code* requires adjusting coding style a bit, to properly decouple concerns and dependencies.

**Comment by Hosch250** [comment]
Why don't we just expose our .NET IMessageBox?

**Comment by retailcoder** [comment]
Because that would be a *managed* WinForms MessageBox, and would break without a reference to our tlb. And a VBA class would have educational purposes too =)

**Comment by Hosch250** [comment]
OK, but we should not just automatically add it to each project--it should be a command of its own, like the "Test Module" command.

**Comment by SystemsModelling** [comment]
Or simply provide a wrapper for the Windows API MessageBoxW which would have the advantage of handling Unicode characters.

Public Declare PtrSafe Function MessageBoxW Lib "user32" _
   (Optional ByVal hwnd As LongPtr, Optional ByVal Prompt As LongPtr, _
   Optional ByVal Caption As LongPtr, Optional ByVal BUTTONS As Long) As Long

Function ShowMsgBox(sPrompt As String, lButtons As VbMsgBoxStyle, sTitle As String) As VbMsgBoxResult

   ShowMsgBox = MessageBoxW(, StrPtr(sPrompt), StrPtr(sTitle), lButtons)

End Function

**Comment by daFreeMan** [comment]
At a previous position I wrote a `MyMsgBox` that had `MsgBox`-like functionality with a built-in countdown timer. 

We needed to be able to boot people out of the system, and many times people would go to lunch with the Access front-end open. A `MsgBox` displayed on their screen would disable the timer event that would check for the "boot 'em out" flag in the DB.

**TL;DR** I support this feature!

**Comment by retailcoder** [comment]
@daFreeMan this issue is made entirely moot with the recent merging of @comintern's work on the fakes framework.

`MsgBox` calls are suddenly no longer problematic as far as unit testing is concerned - the test setup code just needs to configure `Fakes.MsgBox` as needed.
